pykern/fconf.py

```diff
@@ -0,0 +1,569 @@
+            return self
+        return self.__func(name, KeyError)
+
+
+class _XPath:
+    def __init__(self):
+        self.stack = []
+
+    def push(self, key):
+        f = inspect.currentframe().f_back.f_back.f_back
+        self.stack.append(
+            PKDict(key=key, func=f.f_code.co_name, line=f.f_lineno),
+        )
+
+    def pop(self):
+        self.stack.pop()
+
+    def top(self):
+        if self.stack:
+            return self._str(self.stack[-1])
+        else:
+            "None"
+
+    def __str__(self):
+        return "/".join([k.key for k in self.xpath.stack])
+
+    def stack_as_str(self):
+        res = "Evaluator stack:\n"
+        for i in self.stack:
+            res += self._str(i)
+        return res
+
+    def _str(self, element):
+        return f"{element.func}:{element.line} {str(element.key):.100}\n"
+
+
+class _YAMLMacro(PKDict):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        # TEST: duplicate params, invalid params, etc.
+        s = set()
+        for p in self.params:
+            if p in s:
+                raise TypeError(f"duplicate param={p} in macro={self.name}")
+            s.add(p)
+
+    def call(self, namespace, args, kwargs):
+        return namespace._evaluator.start(
+            base=None,
+            local_fvars=self._parse_args(args, kwargs),
+            source=self,
+            xpath_element=self.name + "()",
+        )
+
+    def _kind(self):
+        return "macro"
+
+    def _parse_args(self, args, kwargs):
+        p = list(self.params)
+        res = PKDict({k: _NO_PARAM for k in p})
+        # DOC: there are no optional params in text_macros
+        for a in args:
+            if not p:
+                raise TypeError(f"too many args={args} macro={self.name}")
+            res[p.pop(0)] = a
+        for k, v in kwargs.items():
+            if k not in p:
+                if k not in self.params:
+                    raise TypeError(f"invalid kwarg={k} macro={self.name}")
+                raise TypeError(f"position arg followed by kwarg={k} macro={self.name}")
+            if res[k] is not _NO_PARAM:
+                raise TypeError(f"duplicate kwarg={k} macro={self.name}")
+            res[k] = v
+        x = [k for k, v in res.items() if v is _NO_PARAM]
+        if x:
+            raise TypeError(f"missing args={x} macro={self.name}")
+        return res
+
+    def __str__(self):
+        return f"macro={self.name}"
+
+
+def _exception_reason(exc, reason):
+    reason = "; " + reason
+    if hasattr(exc, "reason"):
+        exc.reason += reason
+    if hasattr(exc, "args"):
+        if exc.args is None:
+            exc.args = tuple()
+        if isinstance(exc.args, (tuple, list)):
+            if len(exc.args) == 0:
+                exc.args = (reason,)
+            elif isinstance(exc.args[0], str):
+                x = list(exc.args)
+                x[0] += reason
+                exc.args = tuple(x)
+    # Other cases?
```

Comment:
Can you add a comment saying this doesn't have a default case

Reply:
4766c22